In [32]:
import numpy as np
import scipy.stats as stats
import pandas as pd

ЗАДАЧА №1. Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy (полученные значения должны быть равны). Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков, а затем с использованием функций из библиотек numpy и pandas.

In [33]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [34]:
# Для расчета ковариации "элементарными действиями" будем использовать формулу cov_xy=μ(XY) − μ(X) ∗ μ(Y):
cov = np.mean(zp * ks) - np.mean(zp) * np.mean(ks)
cov

9157.839999999997

In [35]:
# Теперь посчитаем ковариацию через функцию:
cov_check = np.cov(zp,ks)
cov_check

array([[ 3882.93333333, 10175.37777778],
       [10175.37777778, 33854.32222222]])

In [36]:
# Ой. Но, чтобы выполнить условие задачи и посчитать их одинаковыми, тогда используем смещенность:
cov_check = np.cov(zp,ks,ddof=0)
cov_check

array([[ 3494.64,  9157.84],
       [ 9157.84, 30468.89]])

In [37]:
# Будем использовать смещенное значение ковариации. Это значит, что среднеквадратичные отклонения тоже будем считать смещенными.
σ_zp = np.std(zp,ddof=0)
σ_ks = np.std(ks,ddof=0)
σ_zp, σ_ks

(59.115480206118605, 174.55340157098058)

In [38]:
# Посчитаем коэффициент корреляции Пирсона по формуле:
r_zp_ks = cov / (σ_zp * σ_ks)
r_zp_ks

0.8874900920739158

В тексте задачи не сформулирован такой вопрос, но стоит заметить, что между двумя выборками имеется сильная прямая линейная зависимость.

In [39]:
# Теперь посчитаем с использованием функций из библиотек numpy ...:
r_zp_ks_check = np.corrcoef(zp,ks)
r_zp_ks_check

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

In [40]:
# ... и pandas:
df = pd.DataFrame({'zp': [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
                   'ks': [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]})
df.corr()

zp       ks
zp  1.00000  0.88749
ks  0.88749  1.00000

ОТВЕТ ЗАДАЧИ №1: Значение ковариации - 9157, коэффициент корреляции Пирсона - 0,887.

ЗАДАЧА №2. Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [13]:
x = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
n = len(x)
# Поскольку объем выборки меньше 100, то считаем с условием несмещенности:
σ = np.std(x, ddof=1)
μ = np.mean(x)
n, σ, μ 

(10, 10.54566788359614, 118.1)

In [14]:
# Сигма генеральной совокупности не известна, поэтому применяем критерий Стьюдента:
t = stats.t.ppf(0.975, (n - 1))
t

2.2621571627409915

In [15]:
# Определяем интервалы:
left = μ - t * (σ/np.sqrt(n))
right = μ + t * (σ/np.sqrt(n))
left, right

(110.55608365158724, 125.64391634841274)

ОТВЕТ ЗАДАЧИ №2: Доверительный интервал для математического ожидания IQ: от 111 до 126.

ЗАДАЧА №3. Известно, что рост футболистов в сборной распределен нормально с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27, среднее выборочное составляет 174.2. Найдите доверительный интервал для математического ожидания с надежностью 0.95.

In [16]:
# Из условий задачи знаем:
D_x = 25
σ = np.sqrt(D_x)
μ = 174.2
n = 27
σ

5.0

In [17]:
# Сигма генеральной совокупности известна, поэтому используем z-критерий:
Z = stats.norm.ppf(0.975)
Z

1.959963984540054

In [18]:
# Определяем интервалы:
left = μ - Z * (σ/np.sqrt(n))
right = μ + Z * (σ/np.sqrt(n))
left, right

(172.3140237765397, 176.08597622346028)

ОТВЕТ ЗАДАЧИ №3: Доверительный интервал для математического ожидания IQ: от 172,3 до 176,1.